resolvers/deno/npm/managed/mod.rs

```diff
@@ -8,42 +8,168 @@ mod resolution;
 use std::path::Path;
 use std::path::PathBuf;
 
+use deno_npm::resolution::PackageReqNotFoundError;
+use deno_npm::NpmPackageCacheFolderId;
+use deno_npm::NpmPackageId;
+use deno_path_util::fs::canonicalize_path_maybe_not_exists;
+use deno_semver::package::PackageReq;
 use node_resolver::InNpmPackageChecker;
+use node_resolver::NpmPackageFolderResolver;
 use sys_traits::FsCanonicalize;
 use sys_traits::FsMetadata;
 use url::Url;
 
-pub use self::common::NpmPackageFsResolver;
-pub use self::common::NpmPackageFsResolverPackageFolderError;
+use self::common::NpmPackageFsResolver;
 use self::common::NpmPackageFsResolverRc;
 use self::global::GlobalNpmPackageResolver;
 use self::local::LocalNpmPackageResolver;
 pub use self::resolution::NpmResolution;
-use self::resolution::NpmResolutionRc;
+pub use self::resolution::NpmResolutionRc;
 use crate::sync::new_rc;
+use crate::sync::MaybeSend;
+use crate::sync::MaybeSync;
 use crate::NpmCacheDirRc;
 use crate::ResolvedNpmRcRc;
 
-pub fn create_npm_fs_resolver<
-  TSys: FsCanonicalize + FsMetadata + std::fmt::Debug + Send + Sync + 'static,
->(
-  npm_cache_dir: &NpmCacheDirRc,
-  npm_rc: &ResolvedNpmRcRc,
+#[derive(Debug, thiserror::Error, deno_error::JsError)]
+#[class(generic)]
+#[error("Package folder not found for '{0}'")]
+pub struct NpmManagedResolverPackageFolderError(deno_semver::StackString);
+
+#[derive(Debug, thiserror::Error, deno_error::JsError)]
+pub enum ResolvePkgFolderFromPkgIdError {
+  #[class(inherit)]
+  #[error(transparent)]
+  NpmManagedResolverPackageFolder(#[from] NpmManagedResolverPackageFolderError),
+  #[class(inherit)]
+  #[error(transparent)]
+  Io(#[from] std::io::Error),
+}
+
+#[derive(Debug, thiserror::Error, deno_error::JsError)]
+pub enum ManagedResolvePkgFolderFromDenoReqError {
+  #[class(inherit)]
+  #[error(transparent)]
+  PackageReqNotFound(#[from] PackageReqNotFoundError),
+  #[class(inherit)]
+  #[error(transparent)]
+  ResolvePkgFolderFromPkgId(#[from] ResolvePkgFolderFromPkgIdError),
+}
+
+#[allow(clippy::disallowed_types)]
+pub type ManagedNpmResolverRc<TSys> =
+  crate::sync::MaybeArc<ManagedNpmResolver<TSys>>;
+
+#[derive(Debug)]
+pub struct ManagedNpmResolver<TSys: FsCanonicalize> {
+  fs_resolver: NpmPackageFsResolverRc,
   resolution: NpmResolutionRc,
   sys: TSys,
-  maybe_node_modules_path: Option<PathBuf>,
-) -> NpmPackageFsResolverRc {
-  match maybe_node_modules_path {
-    Some(node_modules_folder) => new_rc(LocalNpmPackageResolver::new(
-      resolution,
+}
```

Comment:
This struct is new. More stuff from the CLI will be moved down here in the future.